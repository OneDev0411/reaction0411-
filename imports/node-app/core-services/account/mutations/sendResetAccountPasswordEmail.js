import _ from "lodash";
import SimpleSchema from "simpl-schema";
import Logger from "@reactioncommerce/logger";
import ReactionError from "@reactioncommerce/reaction-error";
import SimpleSchema from "simpl-schema";
import generateVerificationTokenObject from "@reactioncommerce/api-utils/generateVerificationTokenObject.js";
import Logger from "@reactioncommerce/logger";
import ReactionError from "@reactioncommerce/reaction-error";

const inputSchema = new SimpleSchema({
  email: String
});

/**
 * @method sendResetEmail
 * @memberof Core
 * @summary Send an email with a link that the user can use to reset their password.
 * @param {Object} context - GraphQL execution context
 * @param {Object} account - account object that is related to email address
 * @param {String} email - email of account to reset
 * @returns {Job} - returns a sendEmail Job instance
 */
async function sendResetEmail(context, account, email) {
  const { collections } = context;
  const { Shops, users } = collections;
  // Make sure the user exists, and email is one of their addresses.
  const user = await users.findOne({ _id: account.userId });

  if (!user) {
    Logger.error("sendResetAccountPasswordEmail - User not found");
    throw new ReactionError("not-found", "User not found");
  }

  // make sure we have a valid email
  if (!email || !user.emails || !user.emails.map((mailInfo) => mailInfo.address).includes(email)) {
    Logger.error("sendResetPasswordEmail - Email not found");
    throw new ReactionError("not-found", "Email not found");
  }

  // Create token for password reset
  const tokenObj = generateVerificationTokenObject({ email });

  const { value: updatedAccount } = await users.findOneAndUpdate({ _id: account.userId }, {
    $set: {
      "services.password.reset": tokenObj
    }
  }, {
    returnOriginal: false
  });

  if (!updatedAccount) {
    throw new ReactionError("error-occurred", "Unable to set password reset token");
  }

  // Get shop data for email display
  const shop = await Shops.findOne({ _id: account.shopId });

  const copyrightDate = new Date().getFullYear();

  const dataForEmail = {
    // Shop Data
    shop,
    contactEmail: shop.emails[0].address,
    homepage: context.getAbsoluteUrl(),
    copyrightDate,
    legalName: _.get(shop, "addressBook[0].company"),
    physicalAddress: {
      address: `${_.get(shop, "addressBook[0].address1")} ${_.get(shop, "addressBook[0].address2")}`,
      city: _.get(shop, "addressBook[0].city"),
      region: _.get(shop, "addressBook[0].region"),
      postal: _.get(shop, "addressBook[0].postal")
    },
    shopName: shop.name,
    socialLinks: {
      display: true,
      facebook: {
        display: true,
        icon: `${context.getAbsoluteUrl()}resources/email-templates/facebook-icon.png`,
        link: "https://www.facebook.com"
      },
      googlePlus: {
        display: true,
        icon: `${context.getAbsoluteUrl()}resources/email-templates/google-plus-icon.png`,
        link: "https://plus.google.com"
      },
      twitter: {
        display: true,
        icon: `${context.getAbsoluteUrl()}resources/email-templates/twitter-icon.png`,
        link: "https://www.twitter.com"
      }
    },
    // Account Data
    passwordResetUrl: context.getAbsoluteUrl(`reset-password/${tokenObj.token}`),
    user
  };

  // get account profile language for email
  const language = account.profile && account.profile.language;

  return context.mutations.sendEmail(context, {
    data: dataForEmail,
    fromShop: shop,
    language,
    templateName: "accounts/resetPassword",
    to: email
  });
}

/**
 * @name accounts/sendResetAccountPasswordEmail
 * @summary Checks to see if a user exists for a given email, and sends a password password if user exists
 * @param {Object} context - GraphQL execution context
 * @param {Object} input - Necessary input for mutation. See SimpleSchema.
 * @param {String} input.email - email of account to reset
 * @return {Promise<Object>} with email address if found
 */
export default async function sendResetAccountPasswordEmail(context, input) {
  inputSchema.validate(input);
  const { collections } = context;
  const { Accounts } = collections;
  const {
    email
  } = input;

  const account = await Accounts.findOne({ "emails.address": email });

  if (!account) {
    Logger.error("sendResetAccountPasswordEmail - Account not found");
    throw new ReactionError("not-found", "Account not found");
  }

  const caseInsensitiveEmail = email.toLowerCase();

  await sendResetEmail(context, account, caseInsensitiveEmail);

  return email;
}
