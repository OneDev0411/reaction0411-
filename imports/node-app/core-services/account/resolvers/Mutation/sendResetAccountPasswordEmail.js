/**
 * @name Mutation/sendResetAccountPasswordEmail
 * @summary resolver for the sendResetAccountPasswordEmail GraphQL mutation
 * @param {Object} _ - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {String} args.input.email - email of account to reset
 * @param {String} [args.input.clientMutationId] - An optional string identifying the mutation call
 * @param {Object} context - an object containing the per-request state
 * @returns {Object} r=sendResetAccountPasswordEmailPayload
 */
export default async function sendResetAccountPasswordEmail(_, { input }, context) {
  const { email, clientMutationId = null } = input;

  const emailAddress = await context.mutations.sendResetAccountPasswordEmail(context, {
    email
  });

  return {
    email: emailAddress,
    clientMutationId
  };
}
