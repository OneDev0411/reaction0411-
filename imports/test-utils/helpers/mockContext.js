const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getAbsoluteUrl: jest.fn().mockName("getAbsoluteUrl").mockImplementation((path) => {
    const adjustedPath = path[0] === "/" ? path : `/${path}`;
    return `https://app.mock${adjustedPath}`;
  }),
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  mutations: {},
  queries: {},
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

/**
 * @summary Returns a mock collection instance with the given name
 * @param {String} collectionName The collection name
 * @returns {Object} Mock collection instance
 */
export function mockCollection(collectionName) {
  return {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    bulkWrite: jest.fn().mockName(`${collectionName}.bulkWrite`).mockReturnValue(Promise.resolve({
      nMatched: 2,
      nModified: 2,
      result: {
        writeErrors: []
      }
    })),
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    findOneAndDelete: jest.fn().mockName(`${collectionName}.findOneAndDelete`),
    findOneAndUpdate: jest.fn().mockName(`${collectionName}.findOneAndUpdate`),
    fetch: jest.fn().mockName(`${collectionName}.fetch`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
}

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "ExampleIOUPaymentRefunds",
  "Groups",
  "MediaRecords",
  "NavigationItems",
  "NavigationTrees",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "SimpleInventory",
  "Tags",
  "Templates",
  "Themes",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = mockCollection(collectionName);
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
