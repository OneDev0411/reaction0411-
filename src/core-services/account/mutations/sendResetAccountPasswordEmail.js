import _ from "lodash";
import SimpleSchema from "simpl-schema";
import ReactionError from "@reactioncommerce/reaction-error";
import config from "../config.js";

const { REACTION_IDENTITY_PUBLIC_PASSWORD_RESET_URL } = config;

const inputSchema = new SimpleSchema({
  email: String
});

/**
 * @method sendResetEmail
 * @memberof Core
 * @summary Send an email with a link that the user can use to reset their password.
 * @param {Object} context - GraphQL execution context
 * @param {Object} account - account object that is related to email address
 * @param {String} email - email of account to reset
 * @returns {Job} - returns a sendEmail Job instance
 */
async function sendResetEmail(context, account, email) {
  const {
    collections: { Shops },
    mutations: { startIdentityPasswordReset }
  } = context;

  if (typeof startIdentityPasswordReset !== "function") {
    throw new ReactionError("not-supported", "Password reset not supported");
  }

  const { token } = await startIdentityPasswordReset(context, {
    email,
    userId: account.userId
  });

  // Fall back to primary shop if account has no shop linked
  let shop;
  if (account.shopId) {
    shop = await Shops.findOne({ _id: account.shopId });
  } else {
    shop = await Shops.findOne({ shopType: "primary" });
  }

  if (!shop) throw new ReactionError("not-found", "Shop not found");

  const contactEmail = shop.emails && shop.emails[0] && shop.emails[0].address;

  const dataForEmail = {
    contactEmail,
    homepage: _.get(shop, "storefrontUrls.storefrontHomeUrl", null),
    copyrightDate: new Date().getFullYear(),
    legalName: _.get(shop, "addressBook[0].company"),
    physicalAddress: {
      address: `${_.get(shop, "addressBook[0].address1")} ${_.get(shop, "addressBook[0].address2")}`,
      city: _.get(shop, "addressBook[0].city"),
      region: _.get(shop, "addressBook[0].region"),
      postal: _.get(shop, "addressBook[0].postal")
    },
    shopName: shop.name,
    // Account Data
    passwordResetUrl: REACTION_IDENTITY_PUBLIC_PASSWORD_RESET_URL.replace("TOKEN", token)
  };

  // get account profile language for email
  const language = account.profile && account.profile.language;

  return context.mutations.sendEmail(context, {
    data: dataForEmail,
    fromShop: shop,
    language,
    templateName: "accounts/resetPassword",
    to: email
  });
}

/**
 * @name accounts/sendResetAccountPasswordEmail
 * @summary Checks to see if a user exists for a given email, and sends a password password if user exists
 * @param {Object} context - GraphQL execution context
 * @param {Object} input - Necessary input for mutation. See SimpleSchema.
 * @param {String} input.email - email of account to reset
 * @return {Promise<Object>} with email address if found
 */
export default async function sendResetAccountPasswordEmail(context, input) {
  inputSchema.validate(input);
  const { collections } = context;
  const { Accounts } = collections;
  const {
    email
  } = input;

  const caseInsensitiveEmail = email.toLowerCase();

  const account = await Accounts.findOne({ "emails.address": caseInsensitiveEmail });
  if (!account) throw new ReactionError("not-found", "Account not found");

  await sendResetEmail(context, account, caseInsensitiveEmail);

  return email;
}
