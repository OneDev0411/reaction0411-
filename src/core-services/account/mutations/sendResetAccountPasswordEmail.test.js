import mockContext from "@reactioncommerce/api-utils/tests/mockContext.js";
import sendResetAccountPasswordEmail from "./sendResetAccountPasswordEmail.js";

mockContext.mutations.sendResetAccountPasswordEmail = jest.fn().mockName("mutations.sendResetAccountPasswordEmail");

test("throws if account matching email not found", async () => {
  mockContext.collections.Accounts.findOne.mockReturnValueOnce(undefined);
  await expect(sendResetAccountPasswordEmail(mockContext, { email: "test@email.com" })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if account matching email is found, but shop is not found", async () => {
  mockContext.collections.Accounts.findOne.mockReturnValueOnce({ _id: "123", emails: [{ address: "test@email.com" }], userId: "654" });
  mockContext.mutations.startIdentityPasswordReset = jest.fn().mockName("startIdentityPasswordReset").mockReturnValueOnce({ token: "TOKEN" });
  await expect(sendResetAccountPasswordEmail(mockContext, { email: "test@email.com" })).rejects.toThrowErrorMatchingSnapshot();
  expect(mockContext.mutations.startIdentityPasswordReset).toHaveBeenCalledWith(mockContext, {
    email: "test@email.com",
    userId: "654"
  });
});
