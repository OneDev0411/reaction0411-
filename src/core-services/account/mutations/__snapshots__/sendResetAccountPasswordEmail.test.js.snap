// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if account matching email is found, but shop is not found 1`] = `"Shop not found"`;

exports[`throws if account matching email not found 1`] = `"Account not found"`;
